query-primitives/src/main/scala/io/shiftleft/queryprimitives/steps/types/expressions/ControlStructure.scala

```diff
@@ -25,8 +25,23 @@ class ControlStructure[Labels <: HList](raw: GremlinScala.Aux[nodes.ControlStruc
   /**
     * The expression introduced by this control structure, if any
     * */
-  def condition: Expression[Labels] =
-    new Expression(raw.out.has(NodeKeys.ORDER, new Integer(1)).cast[nodes.Expression])
+  def condition: Expression[Labels] = {
+    new Expression(
+      // I wanted to use flatmap here instead of
+      // map ... filterOnEnd(_.isDefined).map(_.get)
+      // not quite sure why it doesn't work.
+      map { node =>
+        if (isForLoop(node)) {
+          node.start.astChildren.order(2).headOption
+        } else {
+          node.start.astChildren.order(1).headOption
+        }
+      }.filterOnEnd(_.isDefined).map(_.get).raw.cast[nodes.Expression]
+    )
+  }
+
+  private def isForLoop(node: nodes.AstNode): Boolean =
+    node.start.astChildren.l.size == 4
```

Comment:
Necessarily? You can have `for` loops in some languages with only the body for example.

Reply:
This currently only works for the fuzzy C frontend. We are missing a node of type "CONDITION" here because without it, there's no way of knowing exactly which of the expressions is the condition.

Reply:
Just talked to @ml86 about this. We'll introduce a CONDITION edge, to be drawn between the control structure and the condition.